### PR TITLE
Ftpm2

### PR DESCRIPTION
**Pull request by jenswi-linaro**
This PR replaces https://github.com/OP-TEE/build/pull/780

### CODE REVIEW

File: common.mk [hunk 1]
```diff
@@ -556,6 +557,7 @@ endif
 
 CFG_IN_TREE_EARLY_TAS ?= trusted_keys/f04a0fe7-1f5d-4b9b-abf7-619b85b4ce8c
 
+
```
Comment (etienne-lms) [inline common.mk:560]: can discard this change

File: common.mk [hunk 2]
```diff
@@ -582,26 +588,22 @@ optee-os-devkit:
...
-	$(FTPM_FLAGS) $(MAKE) -C $(FTPM_PATH)
+	$(FTPM_FLAGS) $(MAKE) -C $(OPTEE_FTPM_PATH)
 endif
 
 .PHONY: ftpm-clean
 ftpm-clean:
 ifeq ($(MEASURED_BOOT_FTPM),y)
 ftpm-clean:
-	-$(FTPM_FLAGS) $(MAKE) -C $(FTPM_PATH) clean
+	-$(FTPM_FLAGS) $(MAKE) -C $(OPTEE_FTPM_PATH)
```
Comment (etienne-lms) [inline common.mk:608]: restore `clean` target.

File: common.mk [hunk 3]
```diff
@@ -566,6 +568,10 @@ OPTEE_OS_COMMON_FLAGS ?= \
 	DEBUG=$(DEBUG) \
 	CFG_IN_TREE_EARLY_TAS="$(CFG_IN_TREE_EARLY_TAS)"
 
+ifeq ($(MEASURED_BOOT_FTPM),y)
+OPTEE_OS_COMMON_EXTRA_FLAGS += EARLY_TA_PATHS=$(OPTEE_FTPM_PATH)/out/bc50d971-d4c9-42c4-82cb-343fb7f37896.stripped.elf
+optee-os-common: ftpm
```
Comment (etienne-lms) [inline common.mk:573]: Shouldn't `.PHONY: optee-os-common` be set before 1st occurrence of the `optee-os-common` target?

File: common.mk [hunk 4]
```diff
@@ -582,26 +588,22 @@ optee-os-devkit:
...
-# OPTEE_OS_TA_DEV_KIT_DIR variable.
 FTPM_FLAGS ?= 						\
-	TA_CPU=cortex-a9				\
-	TA_CROSS_COMPILE=$(AARCH32_CROSS_COMPILE)	\
-	TA_DEV_KIT_DIR=$(OPTEE_OS_PATH)/out/arm/export-ta_arm32 \
-	CFG_TA_DEBUG=y CFG_TEE_TA_LOG_LEVEL=4 CFG_TA_MEASURED_BOOT=y
+	CROSS_COMPILE=$(CROSS_COMPILE_S_USER)	\
+	TA_DEV_KIT_DIR=$(OPTEE_OS_TA_DEV_KIT_DIR) \
+	CFG_MS_TPM_20_REF=$(MS_TPM_20_REF_PATH) \
+	O=out
```
Comment (etienne-lms) [inline common.mk:595]: As per current https://github.com/OP-TEE/optee_ftpm/blob/master/README.md, `CFG_TA_MEASURED_BOOT` is default disabled. I think it should be enabled here, as previously done. 

Should `CFG_TA_DEBUG` be set when `DEBUG!=0`?

File: qemu_v8.mk [hunk 5]
```diff
@@ -58,6 +61,7 @@ PAUTH ?= n
 # Option to configure Memory Tagging Extension
 MEMTAG ?= n
 
+
```
Comment (etienne-lms) [inline qemu_v8.mk:64]: Remove this extra empty line?

File: common.mk (hunk 3 above)
Comment (jenswi-linaro) [reply common.mk:573]: It doesn't matter (I tested to create an `optee-os-common` file) and the gnu make documentation doesn't mention it explicitly, but it's good practice. I'll move it.

File: common.mk (hunk 4 above)
Comment (jenswi-linaro) [reply common.mk:596]: > As per current https://github.com/OP-TEE/optee_ftpm/blob/master/README.md, `CFG_TA_MEASURED_BOOT` is default disabled. I think it should be enabled here, as previously done.

OK

> 
> Should `CFG_TA_DEBUG` be set when `DEBUG!=0`?

Sure, that makes sense.

**Comment by jenswi-linaro** [comment]
Comments addressed.

**Comment by jenswi-linaro** [comment]
Squashed and rebased

**Comment by jforissier** [comment]
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

Could we have at least a basic "sanity check" test?

**Comment by jenswi-linaro** [comment]
Tag applied.


> Could we have at least a basic "sanity check" test?

Yes. The simplest case I can think of is checking that `/dev/tpm0` is available if the fTPM TA is enabled.
I think it's best to check if the fTPM TA is enabled by trying to open a session to it. I'll add a test in xtest.

**Comment by jforissier** [comment]
> Tag applied.
> 
> > Could we have at least a basic "sanity check" test?
> 
> Yes. The simplest case I can think of is checking that `/dev/tpm0` is available if the fTPM TA is enabled. I think it's best to check if the fTPM TA is enabled by trying to open a session to it. I'll add a test in xtest.

Sounds good!

**Pull request merged**